File: shell-zig/macos/Sources/SkriveShell/CrashLog.swift
import AppKit
import Foundation

// Local, privacy-preserving crash + diagnostics logging (Stage 6.5). Everything
// is written to ~/Library/Application Support/Skrive/crashes and NEVER uploaded
// — the user grabs the folder by hand via Settings → "Reveal diagnostics" and
// sends it in. This matches Skrive's no-telemetry posture and, as a side
// benefit, brings field crashes to the Mac-based dev without a reproduction.
//
// Three native sources are captured here; the renderer error path comes in over
// the `log:append` command (CoreBridge), and webview content-process death
// comes in via the navigation delegate. macOS already writes a full,
// symbolicated .ips report for native crashes under ~/Library/Logs/
// DiagnosticReports — these handlers add a Skrive-owned breadcrumb plus capture
// the things the OS report does NOT cover (uncaught NSExceptions, renderer JS
// errors, webview process death).

// Precomputed C strings for the signal handler. A signal handler may only call
// async-signal-safe functions, so we cannot allocate, format, or touch Swift
// String machinery inside it. These are strdup'd once at install().
// Written once at install() on the main thread before any other thread exists,
// then only read inside the signal handler — externally synchronized by program
// order, so the unchecked annotation is sound.
nonisolated(unsafe) private var crashPathC: UnsafeMutablePointer<CChar>?
nonisolated(unsafe) private var crashMarkerC: UnsafeMutablePointer<CChar>?

// Signal handler: append a fixed marker + a raw backtrace to a pre-resolved
// path using only async-signal-safe calls, then restore the default disposition
// and re-raise so the OS still writes its full crash report. backtrace /
// backtrace_symbols_fd are the standard fd-direct backtrace tools intended for
// exactly this; the buffer lives on the handler's own stack (no malloc).
private func skriveSignalHandler(_ sig: Int32) {
    if let path = crashPathC {
        let fd = open(path, O_WRONLY | O_CREAT | O_APPEND, 0o644)
        if fd >= 0 {
            if let marker = crashMarkerC { _ = write(fd, marker, strlen(marker)) }
            var frames = [UnsafeMutableRawPointer?](repeating: nil, count: 128)
            let count = backtrace(&frames, Int32(frames.count))
            backtrace_symbols_fd(&frames, count, fd)
            close(fd)
        }
    }
    signal(sig, SIG_DFL)
    raise(sig)
}

// Uncaught Objective-C / Cocoa exceptions. Unlike signals this runs in a normal
// context, so Foundation is fair game — write a full, human-readable report.
private func skriveExceptionHandler(_ exception: NSException) {
    let lines = [
        "name: \(exception.name.rawValue)",
        "reason: \(exception.reason ?? "(none)")",
        "userInfo: \(exception.userInfo.map { String(describing: $0) } ?? "(none)")",
        "stack:",
        exception.callStackSymbols.joined(separator: "\n")
    ]
    CrashLog.writeReport(prefix: "exception", body: lines.joined(separator: "\n"))
}

enum CrashLog {
    /// ~/Library/Application Support/Skrive/crashes — the same app-data root the
    /// core uses (Resources.configJSON), with the crashes subfolder.
    static let crashesDir: URL = {
        let appSupport = FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)
            .first ?? URL(fileURLWithPath: NSHomeDirectory())
        return appSupport
            .appendingPathComponent("Skrive", isDirectory: true)
            .appendingPathComponent("crashes", isDirectory: true)
    }()

    /// Install the native crash handlers. Call as early as possible (before the
    /// app runs) so a crash during startup is still captured.
    static func install() {
        ensureDir()

        // Stash the native-crash log path + marker as C strings for the signal
        // handler (which can't allocate). All native crashes append to one file.
        let nativeLog = crashesDir.appendingPathComponent("native-crash.log").path
        crashPathC = strdup(nativeLog)
        crashMarkerC = strdup(
            "\n=== Skrive host crash (signal). Full symbolicated report: "
                + "~/Library/Logs/DiagnosticReports. Backtrace follows ===\n"
        )

        NSSetUncaughtExceptionHandler(skriveExceptionHandler)

        // The fatal signals worth a breadcrumb. SIGABRT also covers Zig core
        // panics (which trap), so a core panic lands here as a native crash.
        for sig in [SIGSEGV, SIGABRT, SIGILL, SIGBUS, SIGFPE, SIGTRAP] {
            signal(sig, skriveSignalHandler)
        }
    }

    /// Append a renderer-diagnostics line (from the `log:append` command). Main
    /// thread; renderer errors are rare, so a plain seek-and-append is fine.
    static func append(_ line: String) {
        ensureDir()
        let url = crashesDir.appendingPathComponent("renderer.log")
        let stamped = "[\(timestamp())] \(line)\n"
        guard let data = stamped.data(using: .utf8) else { return }
        if let handle = try? FileHandle(forWritingTo: url) {
            defer { try? handle.close() }
            _ = try? handle.seekToEnd()
            try? handle.write(contentsOf: data)
        } else {
            try? data.write(to: url)
        }
    }

    /// Webview content-process death — not a host crash (no .ips), so worth its
    /// own breadcrumb. The host reloads the renderer to recover.
    static func logWebviewTermination() {
        writeReport(
            prefix: "webview",
            body: "The WebKit content process terminated; reloading the renderer."
        )
    }

    /// Open the crashes folder in Finder (Settings → "Reveal diagnostics").
    static func reveal() {
        ensureDir()
        NSWorkspace.shared.open(crashesDir)
    }

    /// Write a single timestamped report file (used by the exception + webview
    /// paths, which run in a normal context).
    static func writeReport(prefix: String, body: String) {
        ensureDir()
        let name = "\(prefix)-\(Int(Date().timeIntervalSince1970)).log"
        let url = crashesDir.appendingPathComponent(name)
        let text = "[\(timestamp())] Skrive \(prefix)\n\(body)\n"
        try? text.data(using: .utf8)?.write(to: url)
    }

    private static func ensureDir() {
        try? FileManager.default.createDirectory(
            at: crashesDir, withIntermediateDirectories: true
        )
    }

    private static func timestamp() -> String {
        let f = ISO8601DateFormatter()
        return f.string(from: Date())
    }
}

File: shell-zig/windows/src/crashlog.zig
//! Local, privacy-preserving crash + diagnostics logging (Stage 6.5) — the
//! Windows twin of the macOS host's CrashLog.swift. Everything is written to
//! %APPDATA%\Skrive\crashes and NEVER uploaded; the user grabs the folder by
//! hand via Settings -> "Reveal diagnostics" and sends it in. This matches
//! Skrive's no-telemetry posture and brings field crashes to the (Mac-based)
//! dev without a reproduction.
//!
//! Three native sources land here, mirroring the macOS handlers:
//!   1. Unhandled native exceptions (incl. Zig host/core panics, which @trap to
//!      an illegal-instruction exception) -> SetUnhandledExceptionFilter writes
//!      a minidump (`native-crash.dmp`) + a breadcrumb. This is the exact
//!      parity of the mac signal handler, where a core panic traps to SIGABRT.
//!   2. Renderer JS errors (window.onerror / unhandledrejection) -> the
//!      sandboxed renderer can't write files, so it forwards them over the
//!      `log:append` command and the host appends to `renderer.log`.
//!   3. WebView2 content-process death -> a breadcrumb via the ProcessFailed
//!      handler (app.zig), which then reloads the renderer to recover.
//!
//! Windows already writes its own crash report via WER/DiagnosticReports; the
//! exception filter adds a Skrive-owned minidump + breadcrumb and returns
//! EXCEPTION_CONTINUE_SEARCH so that default WER handling still runs.

const std = @import("std");
const win32 = @import("win32.zig");
const paths = @import("paths.zig");
const host_cmds = @import("host_cmds.zig");

// Wide, NUL-terminated paths resolved once in install() (before any other
// thread exists) and only read inside the exception filter, which must not
// allocate or call the std/libc machinery. Externally synchronized by program
// order — install() runs first on the main thread — so the unprotected globals
// are sound. Allocated with c_allocator and intentionally leaked (process
// lifetime), mirroring the macOS host's strdup'd C strings.
var dump_path_w: ?[*:0]const u16 = null;
var native_log_w: ?[*:0]const u16 = null;

/// Install the native crash handler. Call as early as possible (before the
/// window/webview and any worker thread) so a crash during startup is still
/// captured. Best-effort: a failure to resolve the crashes dir leaves the
/// handler uninstalled rather than aborting startup.
pub fn install(gpa: std.mem.Allocator) void {
    const dir = crashesDir(gpa) catch return;
    defer gpa.free(dir);

    dump_path_w = stashWide(gpa, dir, "native-crash.dmp");
    native_log_w = stashWide(gpa, dir, "native-crash.log");

    _ = win32.SetUnhandledExceptionFilter(crashFilter);
}

/// Top-level exception filter. Async-unsafe work is avoided: only Win32 calls
/// and stack-buffer formatting (no heap, no libc buffering). Writes a minidump
/// + a breadcrumb, then lets default WER handling proceed.
fn crashFilter(info: ?*anyopaque) callconv(win32.WINAPI) win32.LONG {
    if (native_log_w) |lp| {
        var st: win32.SYSTEMTIME = undefined;
        win32.GetLocalTime(&st);
        var buf: [256]u8 = undefined;
        const line = std.fmt.bufPrint(
            &buf,
            "[{d:0>4}-{d:0>2}-{d:0>2}T{d:0>2}:{d:0>2}:{d:0>2}] Skrive host crash (unhandled exception); minidump: native-crash.dmp\n",
            .{ st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond },
        ) catch "";
        appendWide(lp, line);
    }

    if (dump_path_w) |dp| {
        const h = win32.CreateFileW(dp, win32.GENERIC_WRITE, 0, null, win32.CREATE_ALWAYS, win32.FILE_ATTRIBUTE_NORMAL, null);
        if (h != win32.INVALID_HANDLE_VALUE) {
            var exinfo = win32.MINIDUMP_EXCEPTION_INFORMATION{
                .ThreadId = win32.GetCurrentThreadId(),
                .ExceptionPointers = info,
                .ClientPointers = 0,
            };
            _ = win32.MiniDumpWriteDump(
                win32.GetCurrentProcess(),
                win32.GetCurrentProcessId(),
                h,
                win32.MiniDumpNormal,
                &exinfo,
                null,
                null,
            );
            _ = win32.CloseHandle(h);
        }
    }

    return win32.EXCEPTION_CONTINUE_SEARCH;
}

/// Append a renderer-diagnostics line (from `log:append`). Normal UI-thread
/// context, so the allocator is fair game; renderer errors are rare.
pub fn appendRenderer(gpa: std.mem.Allocator, line: []const u8) void {
    const dir = crashesDir(gpa) catch return;
    defer gpa.free(dir);
    const path = std.fs.path.join(gpa, &.{ dir, "renderer.log" }) catch return;
    defer gpa.free(path);
    appendUtf8(gpa, path, "{s} {s}\n", .{ timestamp(), line });
}

/// WebView2 content-process death — not a host crash (no minidump), so it gets
/// its own breadcrumb. The host reloads the renderer to recover (app.zig).
pub fn logWebviewTermination(gpa: std.mem.Allocator) void {
    const dir = crashesDir(gpa) catch return;
    defer gpa.free(dir);
    const path = std.fs.path.join(gpa, &.{ dir, "native-crash.log" }) catch return;
    defer gpa.free(path);
    appendUtf8(gpa, path, "{s} Skrive webview: content process terminated; reloading the renderer.\n", .{timestamp()});
}

/// Open the crashes folder in Explorer (Settings -> "Reveal diagnostics").
pub fn reveal(gpa: std.mem.Allocator) void {
    const dir = crashesDir(gpa) catch return;
    defer gpa.free(dir);
    _ = host_cmds.reveal(gpa, dir);
}

// ---- helpers --------------------------------------------------------------

/// %APPDATA%\Skrive\crashes, created if missing. Caller owns the result.
fn crashesDir(gpa: std.mem.Allocator) ![]u8 {
    const app_data = try paths.appDataDir(gpa);
    defer gpa.free(app_data);
    const dir = try std.fs.path.join(gpa, &.{ app_data, "crashes" });
    errdefer gpa.free(dir);
    const dir_w = try std.unicode.utf8ToUtf16LeAllocZ(gpa, dir);
    defer gpa.free(dir_w);
    _ = win32.CreateDirectoryW(dir_w.ptr, null); // ignore "already exists"
    return dir;
}

/// Build `dir\name`, convert to a wide NUL-terminated string, and leak it for
/// the process lifetime (the exception filter holds the pointer). Returns null
/// on any failure (the filter then skips that artifact).
fn stashWide(gpa: std.mem.Allocator, dir: []const u8, name: []const u8) ?[*:0]const u16 {
    const path = std.fs.path.join(gpa, &.{ dir, name }) catch return null;
    defer gpa.free(path);
    const wide = std.unicode.utf8ToUtf16LeAllocZ(std.heap.c_allocator, path) catch return null;
    return wide.ptr; // intentionally leaked: lives for the whole process.
}

/// Append a UTF-8 line to `path` (created if absent) using raw Win32, so the
/// renderer-log and breadcrumb paths share one mechanism. Best-effort.
fn appendUtf8(gpa: std.mem.Allocator, path: []const u8, comptime fmt: []const u8, args: anytype) void {
    const path_w = std.unicode.utf8ToUtf16LeAllocZ(gpa, path) catch return;
    defer gpa.free(path_w);
    const content = std.fmt.allocPrint(gpa, fmt, args) catch return;
    defer gpa.free(content);
    appendWide(path_w.ptr, content);
}

/// Append raw bytes to the file at `path_w`. Opened with FILE_APPEND_DATA so the
/// write always lands at end-of-file with no seek; safe to call from the
/// exception filter (no allocator, no buffering). Best-effort.
fn appendWide(path_w: [*:0]const u16, bytes: []const u8) void {
    if (bytes.len == 0) return;
    const h = win32.CreateFileW(
        path_w,
        win32.FILE_APPEND_DATA,
        win32.FILE_SHARE_READ | win32.FILE_SHARE_WRITE,
        null,
        win32.OPEN_ALWAYS,
        win32.FILE_ATTRIBUTE_NORMAL,
        null,
    );
    if (h == win32.INVALID_HANDLE_VALUE) return;
    defer _ = win32.CloseHandle(h);
    var written: u32 = 0;
    _ = win32.WriteFile(h, bytes.ptr, @intCast(bytes.len), &written, null);
}

/// ISO-8601-ish local timestamp `[YYYY-MM-DDTHH:MM:SS]` for renderer/webview
/// lines (normal context only — uses a static buffer, single-threaded callers).
fn timestamp() []const u8 {
    const S = struct {
        var buf: [32]u8 = undefined;
    };
    var st: win32.SYSTEMTIME = undefined;
    win32.GetLocalTime(&st);
    return std.fmt.bufPrint(
        &S.buf,
        "[{d:0>4}-{d:0>2}-{d:0>2}T{d:0>2}:{d:0>2}:{d:0>2}]",
        .{ st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond },
    ) catch "[?]";
}
